src/reconcile/rec_row.c

```diff
@@ -1022,6 +1024,16 @@ __wt_rec_row_leaf(
         /* Write any K/V pairs inserted into the page after this key. */
         if ((ins = WT_SKIP_FIRST(WT_ROW_INSERT(page, rip))) != NULL)
             WT_ERR(__rec_row_leaf_insert(session, r, ins));
+        /*
+         * If we're removing a key, we should also remove the history store contents associated with
+         * that key. Even if we fail reconciliation after this point, we're safe to do this. In
+         * order for us to consider removing the key, a globally visible tombstone must exist
+         * meaning that the history store content is obsolete in any case.
+         */
+        if (remove_key && !WT_IS_HS(btree)) {
+            WT_ERR(__wt_row_leaf_key(session, page, rip, tmpkey, true));
```

Comment:
I think it is better to first delete the keys from HS and then from DS

Reply:
As we discussed, this is essentially what is happening. We're in the process of iterating over keys to construct the new page image and as we're doing this, we're deleting history store content.